### ISSUE TITLE
[LAB] Week 2_Problem 2_이성준

### ISSUE THREAD

**Issue by E-pak-sa** [vlab-kaist/NN101_23S#92]
### Problem

Week 2_Problem 2

### Source Code

```python
import torch
from random import random 
from typing import Callable

##                         Problem 2                          ##
##                                                            ##
##            Arbitary x_train, y_train are given.            ##
##     In this problem, x_train is list of list of float.     ##
##   Suppose that x and y have linear correlation, y=wx+b.    ##
##           (In this problem, w will be a vector.)           ##
##     In function training(), you should return [w, b].      ##
##          In function predict(), you should return          ##
##            list y_test corresponding to x_test.            ##
##                  Made by @jangyoujin0917                   ##
##                                                            ##

LR = 0.001
iteration = 200

# NOTE : Feel free to use torch.optim and tensor.
def CostFunction1(X,y,a,b,c):
    cost=0
    for i in range(len(y)):
        cost+=(a*X[i][0]+b*X[i][1]+c-y[i])**2
    return cost/3/len(y)

def Grad_a(X,y,a,b,c,i):
    return 2* ( a*X[0] +b*X[1] +c -y)*X[0]/(i+1) /3
    
def Grad_b(X,y,a,b,c,i):
    return 2* (a*X[0]+b*X[1]+c-y)*X[1] /(i+1) /3

def Grad_c(X,y,a,b,c,i):
    return 2* (a*X[0]+b*X[1]+c-y) /(i+1) / 3

def training(x_train : list[list[float]], y_train : list[float]) -> tuple[list[float], float]: # DO NOT MODIFY FUNCTION NAME
    # data normalization
    # 1. Prevents overflow when calculating MSE
    # 2. Prevents underfitting
    # Note that you need to convert [w, b] to the original scale.
    # w = w * (y_max - y_min)
    # b = b * (y_max - y_min) + y_min
    a,b,c = 0,0,0
    for i in list(range(len(y_train))) * 4:
        for j in range(iteration):
            X,y = x_train[i],y_train[i]
            g_a = Grad_a(X,y,a,b,c,i)
            g_b = Grad_b(X,y,a,b,c,i)
            g_c = Grad_c(X,y,a,b,c,i)
            a -= LR * g_a
            b -= LR * g_b
            c -= LR * g_c
            # if not j %100:
            #     print(a,b,c)
    return [a,b],c
    
    ### IMPLEMENT FROM HERE


def predict(x_train : list[list[float]], y_train : list[float], x_test : list[list[float]]) -> list[float]: # DO NOT MODIFY FUNCTION NAME
    answer = []
    global w,b
    for i in range(len(x_test)):
        [x1,x2] = x_test[i]
        answer.append(x1*w[0] + x2*w[1] + b)
    return answer
    pass ### IMPLEMENT FROM HERE

if __name__ == "__main__":
    x_train = [[0., 1.], [1., 0.], [2., 2.], [3., 1.], [4., 3.],[7., 6.],[1.,1.],[1.,2.],[9.,3.],[-3,-5],[-2,1],[4,-7],[0,-1]]
    y_train = [3., 2., 7., 6., 11., 18.,4.,6.,16,-12,1,-9,-1] # y = x_0 + 2*x_1 + 1 # Note that not all test cases give clear line.
    x_test = [[5., 3.], [10., 6.], [8., 9.]]
    
    w, b = training(x_train, y_train)
    y_test = predict(x_train, y_train, x_test)

    # print(w, b)
    # print(y_test)
```

### Description

gradient descent using partial derivative

**Comment by Dongyeongkim** [comment]
This issue is closed now because of the lack of progress.
